### ISSUE TITLE
Cannot bind IPv6 loopback address inside container

### ISSUE THREAD

**Issue by directhex** [microsoft/azure-pipelines-agent#3276]
vsts-agent creates its own network for container jobs, as per the comments:

https://github.com/microsoft/azure-pipelines-agent/blob/ed9d7ce56cfbf4447d7e77a7e3e52c07b58034d7/src/Agent.Worker/ContainerOperationProvider.cs#L104-L107

The code this equates to is in src/Agent.Worker/Container/DockerCommandManager.cs:

https://github.com/microsoft/azure-pipelines-agent/blob/ed9d7ce56cfbf4447d7e77a7e3e52c07b58034d7/src/Agent.Worker/Container/DockerCommandManager.cs#L220-L232

However, a side effect of this call is it does not explicitly enable IPv6 on the new network, and as a result, there is no IPv6 availability (even just accessing the loopback interface) inside the VM which there would otherwise be if using a correctly configured default network:

```
builder@xam-emag-02:/srv/azdo/a$ /usr/bin/docker network create --label d71aa1 vsts_network_3fa15cd94bfa4326962fd6ee394e83cf
9a77f434d18347035e0fe3017828428ec874429a8aaae00ccb8d1991e24c4e25
builder@xam-emag-02:/srv/azdo/a$ /usr/bin/docker run -it --network vsts_network_3fa15cd94bfa4326962fd6ee394e83cf arm64v8/ubuntu /bin/bash
root@144590f4397c:/# apt update && apt -y install net-tools
[..]
root@144590f4397c:/# ifconfig -a | grep inet6
root@144590f4397c:/#
```

vs

```
builder@xam-emag-02:/srv/azdo/a$ /usr/bin/docker run -it --network bridge arm64v8/ubuntu /bin/bash
root@7e0d7d38019b:/# apt update && apt -y install net-tools
[..]
root@7e0d7d38019b:/# ifconfig -a | grep inet6
        inet6 fe80::42:acff:fe11:2  prefixlen 64  scopeid 0x20<link>
        inet6 2001:db8:1::242:ac11:2  prefixlen 64  scopeid 0x0<global>
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
```

This is preventing us from running some of our test suites on Azure DevOps, e.g. CoreFX tests which explicitly bind the IPv6 loopback address, not the V4 loopback address

I agree with the motivation for using a fresh network for each job run - however, there should be some code to detect whether Docker has been started with IPv6 capability, and pass an `--ipv6` flag appropriately during the `network create`

**Comment by anatolybolshakov** [comment]
Hi @directhex would it work for you if there will be an ability to pass --ipv6 to this command on agent? (e.g. by setting up some pipeline/environment variable)?

**Comment by directhex** [comment]
@anatolybolshakov yeah, trusting the user (e.g. exposing a `network-options:` value which is passed to `docker network create` as well as `options:` passed to `docker create/run`) would put the onus on me, and i'm okay with that!

**Comment by akoeplinger** [comment]
This is not fixed yet, please keep it open.

**Comment by akoeplinger** [comment]
This is not fixed yet, please keep it open.

**Comment by couling** [comment]
I've just bumped into this via AWS ECS.  I have much less control over the networking and to the best of my knowledge creating a network for the "ECS Task" isn't an option.  It uses `bridge`.  I can set it to `host` but I'd prefer not to for security reasons.

It looks like, by default, many linux distributions have an `/etc/hosts` entry:
```
::1         localhost
```
So any call to `bind()` to `localhost` is going to fail.

I've tried adding `/etc/docker/daemon.json`
```
{
  "ipv6": true
}
```
But this has no obvious effect.

Note I'm not trying to enable IPv6 on the docker network, I just want the loopback device to have ip address `::1`.  It seems weird in that in 2022 we're not enabling this by default when most desktops, laptops, servers, IOT devices will have it.

**Comment by couling** [comment]
@anatolybolshakov 

### This is a bug not an enhancement

As far as I can tell docker is supplying `/etc/hosts` which, for example looks like this:
```
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
ff00::0	ip6-mcastprefix
ff02::1	ip6-allnodes
ff02::2	ip6-allrouters
172.17.0.8	00d3337c986f
```

This isn't supplied by the host running docker and it's not supplied by the image... it must be coming from docker (somehow).

Notice that it contains the entry for `::1 localhost`.  So docker is supplying the hostname `localhost` pointing to the address `::1` but then never creating the IP address `::1` on the loopback device.

Docker should either create `::1` or it should NOT supply this entry in `/etc/hosts`.  I think it's rather broken to supply any IPv6 addresses in `/etc/hosts` when there's no IPv6 networking.

**Comment by akoeplinger** [comment]
@anatolybolshakov I see there was a PR that referenced this issue, is that enough to implement this or are we still missing something?

**Comment by ismayilov-ismayil** [comment]
Hi @directhex and @akoeplinger 

We already have this knob, you can use any network related arguments as env variable like --ipv6.
You need to set AZP_AGENT_DOCKER_ADDITIONAL_NETWORK_OPTIONS env variable for this.

Please, note that using --ipv6 without any subnet pool (if is not set on host machine yet) will not take effect, just use it with subnet argument together, like this  ``--ipv6 --subnet=2002:db8:1:1::/64``

Closing this ticket, feel free to ask any other questions or let us know if it's still relevant for you.

**Comment by akoeplinger** [comment]
@ismayilov-ismayil thanks, I tried setting that both as a job-level and global variable but it doesn't work. I can see the `docker network` command in the log is missing my values.

I _think_ this is because the env variables get set _inside_ of the container not outside so they have no effect.

Am I missing something?

**Comment by ismayilov-ismayil** [comment]
@akoeplinger you need to set this variable on Host machine where docker network is created, rather than container inside

**Comment by akoeplinger** [comment]
@ismayilov-ismayil hmm so there's no way to do this with the hosted pool then? that'd be unfortunate
